src/Compilers/Test/Core/SourceGeneration/TestProducers.cs

```diff
@@ -0,0 +1,111 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Immutable;
+using System.IO;
+using System.Text;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Roslyn.Test.Utilities.TestGenerators
+{
+    internal class SingleFileArtifactProducer : IArtifactProducer
+    {
+        private readonly string _content;
+        private readonly string _hintName;
+
+        public SingleFileArtifactProducer(string content, string hintName = "generatedFile")
+        {
+            _content = content;
+            _hintName = hintName;
+        }
+
+        public ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray<DiagnosticDescriptor>.Empty;
+
+        public void Initialize(AnalysisContext context, ArtifactContext artifactContext)
+        {
+            context.RegisterCompilationAction(c => AnalyzeCompilation(c, artifactContext));
```

Comment:
It feels pretty awkward that you have to do this: I'm an artifact producer, yeah I'm going to register something to let me produce artifacts. Can we make this implicit?

Reply:
i don't think we want that.  The point is that the artifact producer wants to decide when they get their callbacks.  They may choose to operate on the compilatoin, or on the semanticmodels/syntaxtrees, or on symbols, etc. etc.  That's why we're using AnalysisContext here so this slots nicely into the whole flexible callback approach the analzyer subsystem already provides.  Except now on top of that, you can generate artifacts while being called back into :)